### PR TITLE
2nd Iter: Ensure that hostnames without dots are excluded.

### LINKED ISSUE #2347: Invalid domain
Hi there,

Thank you for these lists. I notice there is an invalid domain:
0.0.0.0 ttps

On list: https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts

Detection:

  [i] Target: https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
  [✓] Status: Retrieval successful
  [✓] Parsed 188935 exact domains and 0 ABP-style domains (ignored 2 non-domain entries)
      Sample of non-domain entries:
        - "0.0.0.0"
        - "ttps"

Kind regards
Peter

### PR DESCRIPTION
**Pull request by funilrys**
This patch fixes #2347.
This patch touches #2391.

Indeed, my previous patch was missing domains with dashes (-).

### CODE REVIEW

**Comment by funilrys** [comment]
Converting to draft as I caught more errors.

**Comment by funilrys** [comment]
Hey Steve @StevenBlack! Hope you are doing well. 

I have a question because I caught some "discrepancies" while trying to implement this right. 

The question: **What is your position on IPs?**

I'm asking because, on one side, we have:  https://github.com/StevenBlack/hosts/blob/86956052240f198dad3d3a9b74b9c3a4e0f71340/testUpdateHostsFile.py#L841-L853 

which says: we don't want any "raw" IP. 

On the other side, we have: 
https://github.com/StevenBlack/hosts/blob/86956052240f198dad3d3a9b74b9c3a4e0f71340/testUpdateHostsFile.py#L892-L905

which says: if an IP is given as part of a host entry, keep it. 

Finally, on one other side, we have: https://github.com/StevenBlack/hosts/blob/86956052240f198dad3d3a9b74b9c3a4e0f71340/updateHostsFile.py#L1091-L1092

which says: just ignore IPv6. 

That's okay, but we both know, it doesn't make sense at all to have 2 IPv4s in a host's file entry.  

The reason I'm asking is while analyzing the outputs of the coming commits, I discovered, that we have to change the way we catch domains. Indeed, the only records that were not being scrapped correctly, were those where the Punycode of the TLD has a number and dashes in it. That means that to stay as correct as possible, we have to relax the regex - somehow. 
While I did so locally, I now have the problem that IPs pass the regex which was meant for domains. So now I have to change not only the regex but some parts of the code. 

Therefore, we need to decide what we want. 

.... To keep (IPv4s) or not to keep ? That is the question! 😅 


Have a nice day/night!
Nissar

**Comment by StevenBlack** [comment]
Thank you Nissar @funilrys I'll look into the history of that test code now. I agree it doesn't seem to make sense.

**Comment by StevenBlack** [comment]
Nissar @funilrys I'm seeing a failing test here.  I'm still puzzled over the test — what is it actually testing? — it appears to be a test that combines the `--ip` flag with an invalid domain (an ip-address) for input?

Here's what I'm seeing

```bash
 python3 testUpdateHostsFile.py
...........................................F.........................................................................
======================================================================
FAIL: test_no_match (__main__.TestNormalizeRule)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/Steve/Dropbox/dev/hosts/testUpdateHostsFile.py", line 850, in test_no_match
    self.assertEqual(normalize_rule(rule, **kwargs), (None, None))
AssertionError: Tuples differ: ('128.0.0.1', '0.0.0.0 128.0.0.1\n') != (None, None)

First differing element 0:
'128.0.0.1'
None

- ('128.0.0.1', '0.0.0.0 128.0.0.1\n')
+ (None, None)

----------------------------------------------------------------------
Ran 117 tests in 0.091s

FAILED (failures=1)
```

**ALSO** check out some of the warnings above ⬆️ that read like this:



It feels like we are thisclose 🤏🏻 but still not right.

I've taken several stabs at this regex in the past.  I'm currently thinking, regex may not be the best tool anymore at this juncture.

**Comment by buzzingwires** [comment]
So, as I understand, `128.0.0.1` gets matched by `^\s*((?:[\w\-\.]+\.)+[a-zA-Z0-9\-]+)(.*)`, which checks for raw domains not preceded by a destination. It boils down to the question of whether we want to allow IP addresses in the host field, as mentioned by @funilrys

In my opinion, I'm inclined to say no, since mapping IP addresses to other IP addresses is not something the hosts file should be doing.

As for the inefficient regex, I do believe many parsing issues are prevented by not using regexes, but a couple potential optimizations are:

`^\s*(\d{1,3}\.){3}\d{1,3}\s+((?:[\w\-\.]+\.)+[a-zA-Z0-9\-]+)(.*)` -> `^\s*\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\s+([\w\-\.]+\.[a-zA-Z0-9\-]+)(.*)`

Writing out each octet of the IP address seems to reduce the number of steps, at the cost of verbosity. I also removed the first match group `(\d{1,3}\.){3}`, since it didn't seem to be used? I have a strong hunch that `(?:[\w\-\.]+\.)+` is what could potentially be causing the exponential backtracking issue. At the very least, removing the non-capturing group seems to reduce the number of steps. My rationale for safely removing it is that as long as the sequence ends with `.`, preceeding groups should be matched by `[\w\-\.]+`. The effect should be the same as explicitly checking for `.` at the end of each group.

`^\s*((?:[\w\-\.]+\.)+[a-zA-Z0-9\-]+)(.*)` -> `^\s*([\w\-\.]+\.[a-zA-Z0-9\-]+)(.*)`

follows the same ideas. With the `128.0.0.1` test removed, the others fail because the `(.*)` at the end matches the space character and everything after it. Otherwise, might be able to fix this with a lookahead assertion? Assertions could also be used to block IP addresses, if/when relevant.

The output of the hosts file is unchanged, though, save for something that actually appears to be fixed. `0.0.0.0 pgl.example #0101`` -> `0.0.0.0 pgl.example0101`

**Comment by funilrys** [comment]
Hey @StevenBlack ,

I implemented a "regexless" generic solution that should solve all our problems. 
Please look at my last commit and proceed with the tests. 

My last commit assumes we want to get rid of IPs as part of the rule. Therefore, I will need a decision regarding the failing test rules. 

Stay safe and healthy!
Nissar

**Comment by funilrys** [comment]
Following Steve's @StevenBlack comment (CF: https://github.com/funilrys/hosts/commit/b3f93f1528361910abb09e06be0f2ecdc5aeb665#commitcomment-126334510), I submitted the final changes.

Please review. 

Stay safe & healthy!

**Comment by StevenBlack** [comment]
Nissar **THANK YOU SO MUCH** ❤️ 

Tests work great, updating works great.  This is awesome!

MERGING!

**Comment by StevenBlack** [comment]
Thank you for your input as well @buzzingwires that was very valuable too!

**Comment by StevenBlack** [comment]
Nissar @funilrys something I just noticed: domains are no longer all converted to lowercase.

For example, in the base hosts file line `219361`: `PopCash.com`

Also the list jumped by 10k domains with this new version, which is about 8,000 more than I expected.

**Comment by namka279** [comment]
#2347

**Comment by namka279** [comment]
#2391

File: updateHostsFile.py [hunk 1]
```diff
@@ -1061,7 +1061,10 @@
     """
     first try: IP followed by domain
     """
-    regex = r"^\s*(\d{1,3}\.){3}\d{1,3}\s+([\w\.-]+[a-zA-Z])(.*)"
+
+    # WARNING:
+    #   [a-zA-Z0-9\-]+ is NOT an issue. (e.g., xn--p1ai TLD - and others).
+    regex = r"^\s*(\d{1,3}\.){3}\d{1,3}\s+((?:[\w\-\.]+\.)+[a-zA-Z0-9\-]+)(.*)"
```
[the remark this answers is not in the document]
Comment (namka279) [reply updateHostsFile.py]: #2347

File: updateHostsFile.py [hunk 2]
```diff
@@ -1090,7 +1093,9 @@
     """
     # deny any potential IPv6 address here.
     if ":" not in rule:
-        regex = r"^\s*([\w\.-]+[a-zA-Z])(.*)"
+        # WARNING:
+        #   [a-zA-Z0-9\-]+ is NOT an issue. (e.g., xn--p1ai TLD - and others).
+        regex = r"^\s*((?:[\w\-\.]+\.)+[a-zA-Z0-9\-]+)(.*)"
```
[the remark this answers is not in the document]
Comment (namka279) [reply updateHostsFile.py]: #1093

File: updateHostsFile.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (namka279) [reply updateHostsFile.py]: #2400

File: updateHostsFile.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (funilrys) [reply updateHostsFile.py]: Okay @namka279 , please look at the actual code and what has been merged ... This commit didn't land into production ... You are too late to the party ...

@StevenBlack Please close #2634  ...

**Pull request merged**